### ISSUE TITLE
DataCollator inappropriate for Chat Data?

### ISSUE THREAD

**Issue by steremma** [artidoro/qlora#104]
Hello, great paper and repository, thanks. 

I see that the data collator seems to be targeting the instruction tuning use-case where we have an `input` explaining the task and a single `output` solving it. By default we reasonably train only on that output using `IGNORE_INDEX` on the tokens of the `input` part.

However this breaks for chat kind of datasets and since one of the paper's good results is related to OASST1 I wonder whether strong improvements can trivially be achieved by editing the collator. Right now the whole (potentially multi-round) conversation is dumped into the `output` field that the model trains on, nothing is masked by `IGNORE_INDEX`. So training treats the role asking question (user) the same as the role answering (assistant) by training on both. Is that desirable? Why not stick with Vicuna chat-oriented `preprocess` function like [here](https://github.com/lm-sys/FastChat/blob/main/fastchat/train/train.py#L77) which only conditions on the role asking but doesn't back-propagate loss computed oon it?

Apologies if my understanding is incorrect or limited.

**Comment by artidoro** [comment]
Hey Stergiadis, I am glad you find the method interesting!

Your understanding is correct. We are currently training the model on the entire conversation chain including the user queries. 

We conducted initial experiments on the instruction tuning datasets regarding the impact of training on the instruction. We found that on average that causes a performance degradation. See appendix in the paper. 

Based on this observation I agree that there could be some performance improvement by training the model on the response only.

Honestly speaking, we were not expecting the OA dataset to perform this well with such little data so we did not optimize the code for it. If you would like to investigate whether training on the response only improves performance and report back we would be open to update the current repo and model checkpoints.

**Comment by steremma** [comment]
Fair enough, thanks for the quick response. I might find some capacity to investigate and get back to you. Please feel free to close the issue

**Comment by li-aolong** [comment]
@artidoro For instruction tuning, the input is the user's content for single turn. What should  the input be in multi-round conversation? Should be the whole multi-round conversation except the final response? Thank you.
